File: Leetcode/0946 Validate Stack Sequences.ts
function validateStackSequences(pushed: number[], popped: number[]): boolean {
    const pushing: number[] = [];
    let poppedTop = 0;
    for (const val of pushed) {
        pushing.push(val);
        while (
            pushing.length !== 0 
            && poppedTop < popped.length 
            && pushing[pushing.length - 1] === popped[poppedTop]
        ) {
            pushing.pop();
            poppedTop++;
        }
    }
    return poppedTop == popped.length && pushing.length === 0;
}

File: Leetcode/0946 Validate Stack Sequences.py
class Solution:
    def validateStackSequences(self, pushed, popped):
        pushing = []
        popped_top = 0
        for val in pushed:
            pushing.append(val)
            while (
                len(pushing) != 0
                and popped_top < len(popped)
                and pushing[-1] == popped[popped_top]   
            ):
                pushing.pop()
                popped_top += 1
        
        return len(pushing) == 0 and popped_top == len(popped)
